### PR TITLE
DEVENGAGE-2714 link fixes

### PR DESCRIPTION
**Pull request by timsmithgenesys**
This PR makes the following behavior fixes. All changes are expected to be backwards compatible with _valid_ markdown.

- table cell parsing bug fix - escaped pipes (i.e. `\|`) were not being handled correctly
- links with nested square brackets - e.g. `[map[SomeThing]](#SomeThing)`, the fix here is for yeast to support braces within link text. The example link should be modified to be `[map\[SomeThing\]](#SomeThing)`
- link end character parsing now excludes the following characters from being at the end of the parsed URL (they are now left as unparsed text following the URL): `.,!?;`. These are common punctuation used in writing and are often found adjacent to URLs. Despite being valid URL characters, they are not usually intended as such. For URLs that do actually include one of these characters as the last character, use markdown link notation; it does not make this assumption because it has explicit control characters.
- brought bold/italic parsing in line with https://daringfireball.net/projects/markdown/syntax#em, meaning single underscores or asterisks are italics and double are bold. This also adds support for escaped underscores and asterisks.

### CODE REVIEW

File: yeast-markdown-parser/src/__tests__/tests/MarkdownParser.test.ts [hunk 1]
```diff
@@ -491,6 +510,7 @@ test('MarkdownParser parses hyperlink in text', () => {
 
 	// Parse
 	const ast = parser.parse(LINK_MARKDOWN);
+	// debugAST('links', ast);
```
Comment (timsmithgenesys) [inline yeast-markdown-parser/src/__tests__/tests/MarkdownParser.test.ts:513]: I checked in these debug statements as comments on purpose. This is super helpful in debugging tests when running locally and makes it easy for the developer to dump the parsed AST.

File: yeast-markdown-parser/src/plugins/inline/InlineEmphasisPlugin.ts [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+	Token,
+	YeastParser,
+	YeastNodeFactory,
+	YeastInlineNode,
+	YeastInlineNodeTypes,
+	YeastInlineChild,
+} from 'yeast-core';
+
+const ITALICS_REGEX_UNDERSCORES = /(\\?)(_)([^\s_]|\S.*?\S)(\\?)(_)/gi;
+const ITALICS_REGEX_ASTERISKS = /(\\?)(\*)([^\s_]|\S.*?\S)(\\?)(\*)/gi;
```
Comment (timsmithgenesys) [inline yeast-markdown-parser/src/plugins/inline/InlineEmphasisPlugin.ts:12]: I just found an edge case that this isn't handling (a single/unpaired escape within a match). Need to make an update here.

File: yeast-markdown-parser/src/__tests__/tests/MarkdownParser.test.ts [hunk 3]
```diff
@@ -394,24 +384,13 @@ test('MarkdownParser using all inline plugins', () => {
 	// Parse
 	const documentText = fs.readFileSync(path.join(__dirname, '../resources/everythinginline.md'), 'utf8');
 	const ast = parser.parse(documentText);
+	// debugAST('everythinginline', ast);
 
 	// Check document
-	checkAstStructureForDefaultDocument(ast, 1);
```
Comment (timsmithgenesys) [inline yeast-markdown-parser/src/__tests__/tests/MarkdownParser.test.ts:399]: I removed these coded checks in favor of direct/full AST comparison, but made sure all the existing tests were passing before making that change.

**Review by EbenOsei21 — APPROVED** [review]

**Review by diemenschmaschine92 — APPROVED** [review]
